File: SOLUTIONS/LEETCODE/diagonals/sortMatrix/sortMatrix.py
class Solution:
  def sortMatrix(self, grid: List[List[int]]) -> List[List[int]]:
    row = len(grid)          # number of rows
    column = len(grid[0])    # number of columns
    diagonals = {}           # dictionary to hold diagonals by key = i - j

    # Step 1: Collect values from each diagonal
    for i in range(row):
      for j in range(column):
        key = i - j
        if key not in diagonals:
          diagonals[key] = []
        diagonals[key].append(grid[i][j])
    
    # Step 2: Sort each diagonal in descending order
    # (so we can pop from the end later and get smallest first)
    for key in diagonals:
      diagonals[key].sort(reverse = True)
      
    # Step 3: Rewrite grid with sorted values
    for i in range(row):
      for j in range(column):
        key = i - j
        # Pop last element (smallest remaining value)
        grid[i][j] = diagonals[key].pop()
        
    return grid    
          

File: SOLUTIONS/LEETCODE/diagonals/sortMatrix/sortMatrix.js
function sortMatrix(grid) {
  // to track the number of rows and columns 
  let rows = grid.length
  let columns = grid[0].length

  // object (hashmap) to store and monitor diagonals grouped by key = i - j
  let diagonals = {} 

  // Step 1: Collect values from each diagonal
  for (let i = 0; i < rows; i++){
    for (let j = 0; j < columns; j++){
      // Diagonal key is based on row - column index
      const key = i - j

      // Initialize this diagonal if not created yet
      if (!diagonals[key]) diagonals[key] = []
      
      // Push current cell into the correct diagonal bucket
      diagonals[key].push(grid[i][j])
    }
  }

  // Step 2: Sort each diagonal in descending order
  // (we want the largest numbers to come first, so that
  // popping from the end later gives the smallest values first)
  for (let key in diagonals) {
    // declare k as the numerical value of key
    const k = Number(key)
    if (k < 0) {
      // needs ascending when filling -> store descending so pop() gives ascending
      diagonals[key].sort((a, b) => b - a)
    } else {
      // needs descending when filling -> store ascending so pop() gives descending
      diagonals[key].sort((a, b) => a - b)
    }
  }

  // write back using pop()
  for (let i = 0; i < rows; i++){
    for (let j = 0; j < columns; j++){
      // Diagonal key is based on row - column index
      const key = i - j
      grid[i][j] = diagonals[key].pop()
    }
  }

  return grid
}